Doc/library/typing.rst

```diff
@@ -1230,6 +1232,108 @@ These are not used in annotations. They are building blocks for creating generic
     ``covariant=True`` or ``contravariant=True``.  See :pep:`484` for more
     details.  By default, type variables are invariant.
 
+.. class:: TypeVarTuple
+
+    Type variable tuple. A specialised form of :class:`Type variable <TypeVar>`
+    that enables *variadic* generics.
+
+    A normal type variable enables parameterization with a single type. A type
+    variable tuple, in contrast, allows parameterization with an
+    *arbitrary* number of types by acting like an *arbitrary* number of type
+    variables wrapped in a tuple. For example::
+
+        Shape = TypeVarTuple('Shape')
+
+        class Array(Generic[*Shape]): ...
+
+        array_1d: Array[int] = Array()       # 1 type parameter: fine
+        array_2d: Array[int, int] = Array()  # 2 type parameters: also fine
+        array_0d: Array[()] = Array()        # 0 type parameters: also fine
+
+    Note the use of the unpacking operator ``*`` in ``Generic[*Shape]``.
+    Conceptually, you can think of ``Shape`` as a tuple of type variables
+    ``(T1, T2, ...)``. ``Generic[*Shape]`` would then become
+    ``Generic[*(T1, T2, ...)]``, which is equivalent to
+    ``Generic[T1, T2, ...]``. (Note that in older versions of Python, you might
+    see this written using :data:`Unpack <Unpack>` instead, as
+    ``Unpack[Shape]``.)
+
+    Type variable tuples must *always* be unpacked. This helps distinguish type
+    variable types from normal type variables::
+
+        class C(Generic[Shape]): ...  # Not valid (should be Generic[*Shape])
+
+    In cases where you really do just want a tuple of types, make this explicit
+    by wrapping the type variable tuple in ``tuple[]``::
+
+        shape: Shape          # Not valid
+        shape: tuple[Shape]   # Also not valid
+        shape: tuple[*Shape]  # The correct way to do it
+
+    Type variable tuples can be used in the same contexts as normal type
+    variables. For example, in class definitions, arguments, and return types::
+
+        class Array(Generic[*Shape]):
+            def __getitem__(self, key: tuple[*Shape]) -> float: ...
+            def __abs__(self) -> Array[*Shape]: ...
+            def get_shape(self) -> tuple[*Shape]: ...
+
+    Type variable tuples can be happily combined with normal type variables::
+
+        DType = TypeVar('DType')
+
+        class Array(Generic[DType, *Shape]):  # This is fine
+            pass
+
+        class Array2(Generic[*Shape, DType]):  # This would also be fine
+            pass
+
+        float_array_1d: Array[float, Height] = Array()     # Totally fine
+        int_array_2d: Array[int, Height, Width] = Array()  # Yup, fine too
+
+    However, note that at most one type variable tuple may appear in a single
+    list of type arguments or type parameters::
+
+        class Array(Generic[*Shape, *Shape]):  # Not valid
+            pass
```

Comment:
I'd prefer to skip most of these technical details and point to the PEP.

Reply:
(Disagree, see my other comment.)